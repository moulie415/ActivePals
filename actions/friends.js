import * as firebase from "firebase"
export const SET_FRIENDS = 'SET_FRIENDS'
export const REFRESHING_FRIENDS = 'REFRESHING_FRIENDS'


const setFriends = (friends) => ({
	type: SET_FRIENDS,
	friends
})


export const fetchFriends = () => {
	return (dispatch) => {
		let user = firebase.auth().currentUser
		let friends = []
		return firebase.database().ref('users/' + user.uid + '/friends').once('value', snapshot => {

			Object.keys(snapshot.val()).forEach(friend => {
				let promise = new Promise(function(resolve, reject) {
					let status = snapshot.val()[friend]
					firebase.database().ref('users/' + friend).once('value', profile => {
						firebase.storage().ref('images/' + friend ).child('avatar').getDownloadURL() 
						.then(url => {
							resolve({...profile.val(), status, avatar: url})
						})
						.catch(e => {
							resolve({...profile.val(), status})
						})

					})
				})

				friends.push(promise)
			})
		}).then(() => {
			Promise.all(friends).then(items => {
				dispatch(setFriends(items))
			})
		})
	}
}


