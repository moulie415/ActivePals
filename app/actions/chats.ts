import firebase from 'react-native-firebase';
import { MessageType } from '../types/Message';

export const SET_SESSION_CHATS = 'SET_SESSION_CHATS';
export const ADD_SESSION_CHAT = 'ADD_SESSION_CHAT';
export const SET_CHATS = 'SET_CHATS';
export const ADD_CHAT = 'ADD_CHAT';
export const SET_MESSAGE_SESSION = 'SET_MESSAGE_SESSION';
export const NEW_NOTIF = 'NEW_NOTIF';
export const RESET_NOTIFICATION = 'RESET_NOTIFICATION';
export const UPDATE_CHAT = 'UPDATE_CHAT';
export const UPDATE_SESSION_CHAT = 'UPDATE_SESSION_CHAT';
export const SET_GYM_CHAT = 'SET_GYM_CHAT';
export const SET_MESSAGE = 'SET_MESSAGE';
export const RESET_MESSAGE = 'RESET_MESSAGE';
export const SET_UNREAD_COUNT = 'SET_UNREAD_COUNT';
export const MUTE_CHAT = 'MUTE_CHAT';

const setSessionChats = sessionChats => ({
  type: SET_SESSION_CHATS,
  sessionChats,
});

const setChats = chats => ({
  type: SET_CHATS,
  chats,
});

const setMessageSession = (id, messages) => ({
  type: SET_MESSAGE_SESSION,
  id,
  messages,
});

const updateChat = (id, lastMessage) => ({
  type: UPDATE_CHAT,
  id,
  lastMessage,
});

const updateSessionChat = (key, lastMessage) => ({
  type: UPDATE_SESSION_CHAT,
  key,
  lastMessage,
});

const setMute = (id, mute) => ({
  type: MUTE_CHAT,
  id,
  mute,
});

export const setGymChat = chat => ({
  type: SET_GYM_CHAT,
  chat,
});

export const newNotification = notif => ({
  type: NEW_NOTIF,
  notif,
});

export const resetNotification = () => ({
  type: RESET_NOTIFICATION,
});

export const setMessage = (url, text) => ({
  type: SET_MESSAGE,
  url,
  text,
});

export const resetMessage = () => ({
  type: RESET_MESSAGE,
});

export const setUnreadCount = ({ id, count }) => ({
  type: SET_UNREAD_COUNT,
  id,
  count,
});

export const fetchGymChat = gym => {
  return dispatch => {
    return firebase
      .database()
      .ref('gymChats')
      .child(gym)
      .orderByKey()
      .limitToLast(1)
      .on('value', lastMessage => {
        if (lastMessage.val()) {
          const key = Object.keys(lastMessage.val())[0];
          const message = lastMessage.val()[key];
          const chat = { lastMessage: { ...message, key }, key: gym };
          dispatch(setGymChat(chat));
        } else {
          dispatch(setGymChat(null));
        }
      });
  };
};

export const updateLastMessage = notif => {
  return dispatch => {
    if (notif.type === MessageType.MESSAGE) {
      return firebase
        .database()
        .ref('chats')
        .child(notif.chatId)
        .orderByKey()
        .limitToLast(1)
        .once('value', lastMessage => {
          if (lastMessage.val()) {
            const key = Object.keys(lastMessage.val())[0];
            const message = lastMessage.val()[key];
            dispatch(updateChat(notif.uid || notif.friendUid, { ...message, key }));
          }
        });
    }
    if (notif.type === MessageType.SESSION_MESSAGE) {
      return firebase
        .database()
        .ref('sessionChats')
        .child(notif.sessionId)
        .orderByKey()
        .limitToLast(1)
        .once('value', lastMessage => {
          if (lastMessage.val()) {
            const key = Object.keys(lastMessage.val())[0];
            const message = lastMessage.val()[key];

            dispatch(updateSessionChat(notif.sessionId, { ...message, key }));
          }
        });
    }
    if (notif.type === MessageType.GYM_MESSAGE) {
      dispatch(fetchGymChat(notif.gymId));
    }
  };
};

export const resetUnreadCount = id => {
  return (dispatch, getState) => {
    const count = 0;
    const { uid } = getState().profile.profile;
    firebase
      .database()
      .ref(`unreadCount/${uid}`)
      .child(id)
      .set(count);
    dispatch(setUnreadCount({ id, count }));
  };
};

export const getUnreadCount = uid => {
  return (dispatch, getState) => {
    firebase
      .database()
      .ref('unreadCount')
      .child(uid)
      .on('value', snapshot => {
        if (snapshot.val()) {
          Object.keys(snapshot.val()).forEach(id => {
            const count = snapshot.val()[id];
            if (count !== 0) {
              dispatch(setUnreadCount({ id, count }));
            } else if (count !== 0) {
              dispatch(resetUnreadCount(id));
            }
          });
        }
      });
  };
};

export const fetchChats = (uid, limit = 10) => {
  return async dispatch => {
    return firebase
      .database()
      .ref('userChats')
      .child(uid)
      .limitToFirst(limit)
      .on('value', async snapshot => {
        if (snapshot.val()) {
          const chats = await Promise.all(
            Object.keys(snapshot.val()).map(async chat => {
              const chatId = snapshot.val()[chat];
              const lastMessage = await firebase
                .database()
                .ref('chats')
                .child(chatId)
                .orderByKey()
                .limitToLast(1)
                .once('value');
              let message = { text: 'Beginning of chat' };
              if (lastMessage.val()) {
                const key = Object.keys(lastMessage.val())[0];
                message = { ...lastMessage.val()[key], key };
              }
              return { uid: chat, chatId, lastMessage: message, key: chatId };
            })
          );
          const obj = chats.reduce((acc, cur) => {
            acc[cur.uid] = cur;
            return acc;
          }, {});
          dispatch(setChats(obj));
        } else {
          dispatch(setChats({}));
        }
      });
  };
};

export const fetchSessionChats = (uid, limit = 10) => {
  return async dispatch => {
    return firebase
      .database()
      .ref('userSessions')
      .child(uid)
      .limitToFirst(limit)
      .on('value', async userSessions => {
        if (userSessions.val()) {
          const chats = await Promise.all(
            Object.keys(userSessions.val()).map(async session => {
              const type = userSessions.val()[session] === 'private' ? 'privateSessions' : 'sessions';
              const snapshot = await firebase
                .database()
                .ref(`${type}/${session}`)
                .once('value');
              const lastMessage = await firebase
                .database()
                .ref('sessionChats')
                .child(session)
                .orderByKey()
                .limitToLast(1)
                .once('value');
              let message = { text: 'Beginning of chat' };
              if (lastMessage.val()) {
                const key = Object.keys(lastMessage.val())[0];
                message = { ...lastMessage.val()[key], key };
              }
              return { ...snapshot.val(), key: session, lastMessage: message };
            })
          );
          const obj = chats.reduce((acc, cur) => {
            acc[cur.key] = cur;
            return acc;
          }, {});
          dispatch(setSessionChats(obj));
        } else {
          dispatch(setSessionChats({}));
        }
      });
  };
};

export const fetchMessages = (id, amount, uid, endAt) => {
  return async dispatch => {
    const ref = endAt
      ? firebase
          .database()
          .ref(`chats/${id}`)
          .orderByKey()
          .endAt(endAt)
          .limitToLast(amount)
      : firebase
          .database()
          .ref(`chats/${id}`)
          .orderByKey()
          .limitToLast(amount);
    const snapshot = await ref.once('value')
    const messages = {};
    try {
      const url = await firebase
        .storage()
        .ref(`images/${uid}`)
        .child('avatar')
        .getDownloadURL()
      snapshot.forEach(child => {
        if (child.val().user && child.val().user._id === uid) {
          messages[child.key] = {
            ...child.val(),
            key: child.key,
            createdAt: new Date(child.val().createdAt),
            user: { ...child.val().user, avatar: url },
          };
        } else {
          messages[child.key] = { ...child.val(), key: child.key, createdAt: new Date(child.val().createdAt) };
        }
        return false;
      });
      dispatch(setMessageSession(id, messages));
    } catch (e) {
      snapshot.forEach(child => {
        messages[child.key] = { ...child.val(), key: child.key, createdAt: new Date(child.val().createdAt) };
        return false;
      });
      dispatch(setMessageSession(id, messages));
    }
  };
};

export const fetchSessionMessages = (id, amount, isPrivate = false, endAt) => {
  return dispatch => {
    const type = isPrivate ? 'privateSessions' : 'sessions';
    const ref = endAt
      ? firebase
          .database()
          .ref('sessionChats/' + id)
          .endAt(endAt)
          .limitToLast(amount)
      : firebase
          .database()
          .ref('sessionChats/' + id)
          .limitToLast(amount);
    return ref.once('value', snapshot => {
      const messages = {};
      const promises = [];
      firebase
        .database()
        .ref(type + '/' + id)
        .child('users')
        .once('value', users => {
          users.forEach(child => {
            promises.push(
              new Promise(resolve => {
                firebase
                  .storage()
                  .ref('images/' + child.key)
                  .child('avatar')
                  .getDownloadURL()
                  .then(url => resolve({ [child.key]: url }))
                  .catch(e => resolve({ [child.key]: null }));
              })
            );
            return false;
          });
          Promise.all(promises).then(array => {
            let avatars = {};
            array.forEach((avatar, index) => {
              let key = Object.keys(avatar)[0];
              if (key) {
                avatars[key] = avatar[key];
              }
            });
            snapshot.forEach(child => {
              let avatar = child.val().user ? avatars[child.val().user._id] : '';
              if (avatar) {
                messages[child.key] = {
                  ...child.val(),
                  key: child.key,
                  createdAt: new Date(child.val().createdAt),
                  user: { ...child.val().user, avatar },
                };
              } else {
                messages[child.key] = { ...child.val(), key: child.key, createdAt: new Date(child.val().createdAt) };
              }
              return false;
            });
            dispatch(setMessageSession(id, messages));
          });
        });
    });
  };
};

export const fetchGymMessages = (id, amount, endAt) => {
  return dispatch => {
    const ref = endAt
      ? firebase
          .database()
          .ref(`gymChats/${id}`)
          .endAt(endAt)
          .limitToLast(amount)
      : firebase
          .database()
          .ref(`gymChats/${id}`)
          .limitToLast(amount);
    return ref.once('value', snapshot => {
      const messages = {};
      const promises = [];
      firebase
        .database()
        .ref(`gyms/${id}`)
        .child('users')
        .once('value', users => {
          users.forEach(child => {
            promises.push(
              new Promise(resolve => {
                firebase
                  .storage()
                  .ref(`images/${child.key}`)
                  .child('avatar')
                  .getDownloadURL()
                  .then(url => resolve({ [child.key]: url }))
                  .catch(e => resolve({ [child.key]: null }));
              })
            );
            return false;
          });
          Promise.all(promises).then(array => {
            let avatars = {};
            array.forEach((avatar, index) => {
              let key = Object.keys(avatar)[0];
              if (key) {
                avatars[key] = avatar[key];
              }
            });
            snapshot.forEach(child => {
              const avatar = child.val().user ? avatars[child.val().user._id] : '';
              if (avatar) {
                messages[child.key] = {
                  ...child.val(),
                  key: child.key,
                  createdAt: new Date(child.val().createdAt),
                  user: { ...child.val().user, avatar },
                };
              } else {
                messages[child.key] = { ...child.val(), key: child.key, createdAt: new Date(child.val().createdAt) };
              }
              return false;
            });
            dispatch(setMessageSession(id, messages));
          });
        });
    });
  };
};

export const muteChat = (id, mute) => {
  return (dispatch, getState) => {
    const { uid } = getState().profile.profile;
    dispatch(setMute(id, mute));
    firebase
      .database()
      .ref(`muted/${uid}`)
      .child(id)
      .set(mute);
  };
};
