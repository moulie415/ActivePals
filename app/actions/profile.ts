import db from '@react-native-firebase/firestore';
import database from '@react-native-firebase/database';
import auth from '@react-native-firebase/auth';
import storage from '@react-native-firebase/storage';
import {fetchFriends} from './friends';
import {
  fetchSessionChats,
  fetchChats,
  fetchGymChat,
  setGymChat,
  getUnreadCount,
} from './chats';
import {fetchPosts} from './home';
import {fetchSessions, fetchPhotoPath, fetchPrivateSessions} from './sessions';
import Profile, {UserState} from '../types/Profile';
import Place from '../types/Place';
import {MyThunkDispatch, MyThunkResult, Theme} from '../types/Shared';

export const SET_PROFILE = 'SET_PROFILE';
export const SET_LOGGED_IN = 'SET_LOGGED_IN';
export const SET_LOGGED_OUT = 'SET_LOGGED_OUT';
export const SET_GYM = 'SET_GYM';
export const REMOVE_GYM = 'REMOVE_GYM';
export const SET_LOCATION = 'SET_LOCATION';
export const SET_HAS_VIEWED_WELCOME = 'SET_HAS_VIEWED_WELCOME';

interface SetProfileAction {
  type: typeof SET_PROFILE;
  profile: {[key: string]: any};
}

interface SetLoggedOutOutAction {
  type: typeof SET_LOGGED_OUT;
}

interface ViewedWelcomeAction {
  type: typeof SET_HAS_VIEWED_WELCOME;
}

export type ProfileActionTypes =
  | SetProfileAction
  | SetLoggedOutOutAction
  | ViewedWelcomeAction;

export const SetProfile = (profile: {
  [key: string]: any;
}): SetProfileAction => ({
  type: SET_PROFILE,
  profile,
});

export const SetLoggedOut = (): SetLoggedOutOutAction => ({
  type: SET_LOGGED_OUT,
});

export const ViewedWelcome = (): ViewedWelcomeAction => ({
  type: SET_HAS_VIEWED_WELCOME,
});

export const setHasLoggedIn = (loggedIn) => ({
  type: SET_LOGGED_IN,
  loggedIn,
});

export const setLoggedOut = () => ({
  type: SET_LOGGED_OUT,
});

export const setGym = (gym) => ({
  type: SET_GYM,
  gym,
});

export const resetGym = () => ({
  type: REMOVE_GYM,
});

export const setLocation = (location) => ({
  type: SET_LOCATION,
  location,
});

export const setHasViewedWelcome = () => ({
  type: SET_HAS_VIEWED_WELCOME,
});

export const doSetup = (): MyThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    //const {uid} = profile;
    const {uid, gym} = getState().profile.profile;
    try {
      setupPresence(uid);

      //   const fcmToken = await messaging().getToken();
      //   if (fcmToken) {
      //     database()
      //       .ref(`users/${uid}`)
      //       .child('FCMToken')
      //       .set(fcmToken);
      //     console.log(fcmToken);
      //   } else {
      //     console.warn('no token');
      //   }
    } catch (e) {
      console.warn(e);
    }
    dispatch(getUnreadCount(uid));

    dispatch(fetchFriends(uid));

    gym && dispatch(fetchGymChat(gym));
  };
};

const setupPresence = (uid: string) => {
  const ref = database().ref(`users/${uid}`).child('state');
  const lastChange = database().ref(`users/${uid}`).child('lastChange');
  const connectedRef = database().ref('.info/connected');
  connectedRef.on('value', (snap) => {
    if (snap.val() === true) {
      ref.onDisconnect().remove();
      lastChange.onDisconnect().set(database.ServerValue.TIMESTAMP);
      ref.set(UserState.ONLINE);
    }
  });
};

const fetchGym = (profile: Profile): MyThunkResult<Promise<void>> => {
  return async (dispatch: MyThunkDispatch, getState) => {
    if (profile.gym) {
      const snapshot = await database()
        .ref(`gyms/${profile.gym}`)
        .once('value');
      const gym = await fetchPhotoPath(snapshot.val());
      dispatch(setGym(gym));
    }
  };
};

export const fetchProfile = () => {
  return async (dispatch: MyThunkDispatch) => {
    const user = auth().currentUser;
    if (user) {
      const snapshot = await database().ref(`users/${user.uid}`).once('value');
      try {
        const url = await storage()
          .ref(`images/${user.uid}`)
          .child('avatar')
          .getDownloadURL();
        dispatch(SetProfile({...snapshot.val(), avatar: url}));
        dispatch(fetchGym(snapshot.val()));
        return {...snapshot.val(), avatar: url};
      } catch (e) {
        dispatch(SetProfile(snapshot.val()));
        dispatch(fetchGym(snapshot.val()));
        return snapshot.val();
      }
    }
  };
};

export const fetchOther = (uid: string) => {
  return (dispatch: MyThunkDispatch) => {
    dispatch(fetchPrivateSessions());
    dispatch(fetchPosts(uid));
    dispatch(fetchSessions());
    dispatch(fetchSessionChats(uid));
    dispatch(fetchChats(uid));
  };
};

export const removeUser = (): MyThunkResult<Promise<void>> => {
  return async (dispatch: MyThunkDispatch) => {
    const user = auth().currentUser;
    dispatch(setLoggedOut());
    await user?.delete();
  };
};

export const removeGym = (): MyThunkResult<Promise<void>> => {
  return async (dispatch: MyThunkDispatch, getState) => {
    const currentGym = getState().profile.gym.place_id;
    const {uid} = getState().profile.profile;
    database().ref(`users/${uid}`).child('gym').set(null);
    const gym = await database().ref('gyms').child(currentGym).once('value');
    const count = gym.val().userCount - 1;
    database().ref(`gyms/${currentGym}`).child('userCount').set(count);
    database().ref(`gyms/${currentGym}/users`).child(uid).remove();
    dispatch(resetGym());
    dispatch(setGymChat(null));
  };
};

export const joinGym = (location: Place): MyThunkResult<Promise<void>> => {
  return async (dispatch: MyThunkDispatch, getState) => {
    const {uid} = getState().profile.profile;
    if (getState().profile.gym) {
      dispatch(removeGym());
    }
    database().ref(`users/${uid}`).child('gym').set(location.place_id);
    const gym = await database()
      .ref('gyms')
      .child(location.place_id)
      .once('value');
    if (!gym.val()) {
      location.users = {[uid]: true};
      location.userCount = 1;
      database().ref('gyms').child(location.place_id).set(location);
      const systemMessage = {
        _id: 1,
        text: 'Beginning of chat',
        createdAt: new Date().toString(),
        system: true,
      };
      await database().ref(`gymChats/${location.place_id}`).push(systemMessage);
      dispatch(fetchGymChat(location.place_id));
    } else {
      const count = gym.val().userCount ? gym.val().userCount + 1 : 1;
      database()
        .ref(`gyms/${gym.val().place_id}`)
        .child('userCount')
        .set(count);
      database().ref(`gyms/${gym.val().place_id}/users`).child(uid).set(true);
      dispatch(fetchGymChat(location.place_id));
    }
    dispatch(setGym(location));
  };
};
