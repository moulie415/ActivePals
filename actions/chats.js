import * as firebase from "firebase"
import { fetchProfile } from './profile'
export const SET_SESSION_CHATS = 'SET_SESSION_CHATS'
export const ADD_SESSION_CHAT = 'ADD_SESSION_CHAT'
export const REMOVE_SESSION_CHAT = 'REMOVE_SESSION_CHAT'
export const SET_CHATS = 'SET_CHATS'
export const UPDATE_CHATS = 'UPDATE_CHATS'
export const ADD_CHAT = 'ADD_CHAT'
export const SET_MESSAGE_SESSION = 'SET_MESSAGE_SESSION'
export const NEW_NOTIF = 'NEW_NOTIF'
export const RESET_NOTIFICATION = 'RESET_NOTIFICATION'


const setSessionChats = (sessionChats) => ({
	type: SET_SESSION_CHATS,
	sessionChats
})

const addToSessionChats = (session) => ({
	type: ADD_SESSION_CHAT,
	session,
})

export const removeSessionChat = (session) => ({
	type: REMOVE_SESSION_CHAT,
	session
})

const setChats = (chats) => ({
	type: SET_CHATS,
	chats
})

const addToChats = (chat) => ({
	type: ADD_CHAT,
	chat
})

const setMessageSession = (id, messages) => ({
	type: SET_MESSAGE_SESSION,
	id,
	messages,
})

export const updateChats = (chats) => ({
	type: UPDATE_CHATS,
	chats
})

export const newNotification = (notif) => ({
	type: NEW_NOTIF,
	notif
})

export const resetNotification = () => ({
       type: RESET_NOTIFICATION,
})


export const updateLastMessage = (notif) => {
	return (dispatch, getState) => {
		if (notif.type == 'message') {
			return firebase.database().ref('chats').child(notif.chatId).orderByKey().limitToLast(1)
				.once('value', lastMessage => {
					if (lastMessage.val()) {
						let chats = getState().chats.chats.filter(chat => chat.chatId != notif.chatId)
						dispatch(setChats([...chats, {uid: notif.uid, chatId: notif.chatId, lastMessage: Object.values(lastMessage.val())[0]}]))
					}
				})
		}
		else if (notif.type == 'sessionMessage') {
			return firebase.database().ref('sessionChats').child(notif.sessionId).orderByKey().limitToLast(1)
				.once('value', lastMessage => {
					if (lastMessage.val()) {
						let chats = getState().chats.sessionChats.filter(chat => chat.id != notif.sessionId)
						let chat = getState().chats.sessionChats.filter(chat => chat.id == notif.sessionId)[0]
						dispatch(setSessionChats([...chats, {...chat, lastMessage: Object.values(lastMessage.val())[0]}]))
					}
				})

		}
	}
}



export const fetchChats = (chats) => {
	return (dispatch) => {
		let chatList = []
		Object.keys(chats).forEach(chat => {
			let val = chats[chat]
			let promise = new Promise(function(resolve, reject) {
				firebase.database().ref('chats').child(val).orderByKey().limitToLast(1)
				.once('value', lastMessage => {
					let message = {text: "new chat created"}
					if (lastMessage.val()) {
						message = Object.values(lastMessage.val())[0]
					}
					resolve({uid: chat, chatId: val, lastMessage: message})
				})
			})
			chatList.push(promise)
		})
		return Promise.all(chatList).then(chats => {
			dispatch(setChats(chats))
		})
	}
}

export const addChat = (chat) => {
	return (dispatch) => {
		uid = chat.key
		chatId = chat.val()
		return new Promise(resolve => {
			firebase.database().ref('chats/'+ chatId).orderByKey().limitToLast(1)
			.once('value', lastMessage => {
				let message = {text: "new chat created"}
				if (lastMessage.val()) {
					message = Object.values(lastMessage.val())[0]
				}
				resolve({uid, chatId, lastMessage: message})
				dispatch(addToChats({uid, chatId, lastMessage: message}))
				dispatch(fetchProfile())
			})
		})
	}
}

export const fetchSessionChats = (sessions, uid) => {
	return (dispatch) => {
		let chatList = []
		Object.keys(sessions).forEach(session => {
			let promise = new Promise(function(resolve, reject) {
				firebase.database().ref('sessions/' + session).once('value', snapshot => {
					if (snapshot.val()) {
						firebase.database().ref('sessionChats/'+ session).orderByKey().limitToLast(1)
						.once('value', lastMessage => {
							let message = {text: "new session chat created"}
							if (lastMessage.val()) {
								message = Object.values(lastMessage.val())[0]
							}
							resolve({...snapshot.val(), id: session, lastMessage: message})
						})
					}
					else {
						resolve()
						dispatch(removeSessionChat(session))
						firebase.database().ref('users/' + uid + '/sessions').child(session).remove()
					}
				})
			})
			chatList.push(promise)
		})
		return Promise.all(chatList).then(sessionChats => {
			dispatch(setSessionChats(sessionChats))
		})
	}
}

export const addSessionChat = (session) => {
	return (dispatch) => {
		return new Promise(resolve => {
			firebase.database().ref('sessions/' + session).once('value', snapshot => {
				firebase.database().ref('sessionChats/'+ session).orderByKey().limitToLast(1)
				.once('value', lastMessage => {
					let message = {text: "new session chat created"}
					if (lastMessage.val()) {
						message = Object.values(lastMessage.val())[0]
					}
					resolve({...snapshot.val(), id: session, lastMessage: message})
					dispatch(addToSessionChats({...snapshot.val(), id: session, lastMessage: message}))
					dispatch(fetchProfile())
				})
			})
		})
	}
}


export const fetchMessages = (id, amount, uid) => {
	return (dispatch) => {
		return firebase.database().ref('chats/'+ id).orderByKey().limitToLast(amount)
		.once('value', snapshot => {
			let messages = []
			firebase.storage().ref('images/' + uid).child('avatar').getDownloadURL()
			.then (url => {
				snapshot.forEach(child => {
					if (child.val().user && child.val().user._id == uid) {
						messages.push({...child.val(), createdAt: new Date(child.val().createdAt), 
							user: {...child.val().user, avatar: url}})
					}
					else {
						messages.push({...child.val(), createdAt: new Date(child.val().createdAt)})
					}
				})
				dispatch(setMessageSession(id, messages))
			})
			.catch(e => {
				snapshot.forEach(child => {
					messages.push({...child.val(), createdAt: new Date(child.val().createdAt)})
				})
				dispatch(setMessageSession(id, messages))
			})
		})
	}
}

export const fetchSessionMessages = (id, amount) => {
	return (dispatch) => {
		return firebase.database().ref('sessionChats/'+ id).orderByKey().limitToLast(amount)
		.once('value', snapshot => {
			let messages = []
			let promises = []
			firebase.database().ref('sessions/' + id).child('users').once('value', users => {
				users.forEach(child => {
					promises.push(new Promise(resolve => {
						firebase.storage().ref('images/' + child.key ).child('avatar').getDownloadURL()
						.then(url => resolve({[child.key]: url}))
						.catch(e => resolve({[child.key]: null}))
					}))
				})
				Promise.all(promises).then(array => {
					let avatars = {}
					array.forEach((avatar, index) => {
						let key = Object.keys(avatar)[0]
						if (key) {
							avatars[key] = avatar[key]
						}
					})
					snapshot.forEach(child => {
						let avatar = child.val().user? avatars[child.val().user._id] : ""
						if (avatar) {
							messages.push({...child.val(), createdAt: new Date(child.val().createdAt), 
								user: {...child.val().user, avatar}})
						}
						else {
							messages.push({...child.val(), createdAt: new Date(child.val().createdAt)})
							}
					})
					dispatch(setMessageSession(id, messages))
				})
			})
		})
	}
}

