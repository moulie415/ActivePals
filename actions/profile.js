import * as firebase from "firebase"
export const SET_PROFILE = 'SET_PROFILE'
import { fetchFriends } from './friends'
import { fetchSessionChats } from './chats'


const setProfile = (profile) => ({
	type: SET_PROFILE,
	profile
})

export const fetchProfile = () => {
	return (dispatch) => {
		let user = firebase.auth().currentUser
		return firebase.database().ref('users/' + user.uid).once('value', snapshot => {
			firebase.storage().ref('images/' + user.uid ).child('avatar').getDownloadURL() 
			.then(url => {
				dispatch(setProfile({...snapshot.val(), avatar: url}))
			})
			.catch(e => {
				dispatch(setProfile(snapshot.val()))
			})

		})
	}
}

export const doSetup = (profile) => {
	return (dispatch) => {
		return Promise.all([
			profile.val().friends && dispatch(fetchFriends(profile.val().friends)),
			profile.val().sessions && dispatch(fetchSessionChats(profile.val().sessions, profile.val().uid))
			])
	}
}

