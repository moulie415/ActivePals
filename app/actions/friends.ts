import database from '@react-native-firebase/database';
import {upUnreadCount, fetchUsers} from './home';
import Profile, {UserState} from '../types/Profile';
import {fetchOther} from './profile';

export const SET_FRIENDS = 'SET_FRIENDS';
export const SET_FRIEND = 'SET_FRIEND';
export const ADD_FRIEND = 'ADD_FRIEND';
export const UPDATE_FRIEND_STATE = 'UPDATE_FRIEND_STATE';

const setFriends = (friends) => ({
  type: SET_FRIENDS,
  friends,
});

const setFriend = (friend) => ({
  type: SET_FRIEND,
  friend,
});

const addToFriends = (uid, friend) => ({
  type: ADD_FRIEND,
  uid,
  friend,
});

export const updateFriendState = (uid, state) => ({
  type: UPDATE_FRIEND_STATE,
  uid,
  state,
});

export const removeFriend = (uid: string) => {
  return (dispatch, getState) => {
    const {friends}: {[key: string]: Profile} = getState().friends;
    const friendArr = Object.values(friends).filter(
      (friend) => friend.uid !== uid,
    );
    const obj = friendArr.reduce((acc, cur, i) => {
      acc[cur.uid] = cur;
      return acc;
    }, {});
    dispatch(setFriends(obj));
  };
};

const getStateString = (state) => {
  if (state && state !== UserState.AWAY) {
    return UserState.ONLINE;
  }
  if (!state) {
    return UserState.OFFLINE;
  }
  return UserState.AWAY;
};

export const fetchFriends = (uid: string, limit = 10, startAt?: string) => {
  return async (dispatch, getState) => {
    const ref = database().ref('userFriends').child(uid).limitToLast(limit);
    await ref.on('value', async (snapshot) => {
      if (snapshot.val()) {
        const {friends} = getState().friends;
        Object.keys(friends).forEach((friend) => {
          if (!snapshot.val()[friend]) {
            dispatch(removeFriend(friend));
          }
        });
        const uids = Object.keys(snapshot.val());
        await Promise.all(
          uids.map((friendUid) => {
            return new Promise((resolve) => {
              const status = snapshot.val()[friendUid];
              firebase
                .database()
                .ref('users')
                .child(friendUid)
                .on('value', async (profile) => {
                  if (profile.val()) {
                    const {state} = profile.val();
                    const userState = getStateString(state);
                    try {
                      const avatar = await firebase
                        .storage()
                        .ref(`images/${friendUid}`)
                        .child('avatar')
                        .getDownloadURL();
                      dispatch(
                        setFriend({
                          ...profile.val(),
                          state: userState,
                          avatar,
                          status,
                        }),
                      );
                    } catch (e) {
                      dispatch(
                        setFriend({...profile.val(), state: userState, status}),
                      );
                    }
                    resolve();
                  }
                });
            });
          }),
        );
        dispatch(fetchOther(uid));
      } else {
        dispatch(setFriends({}));
        dispatch(fetchOther(uid));
      }
    });
  };
};

export const sendRequest = (friendUid) => {
  return async (dispatch, getState) => {
    const {uid} = getState().profile.profile;
    const promise1 = database()
      .ref(`userFriends/${uid}`)
      .child(friendUid)
      .set('outgoing');
    const promise2 = database()
      .ref(`userFriends/${friendUid}`)
      .child(uid)
      .set('incoming');
    await Promise.all([promise1, promise2]);
    const date = new Date().toString();
    const ref = database().ref('notifications').push();
    const {key} = ref;
    await ref.set({date, uid, type: 'friendRequest'});
    database().ref(`userNotifications/${friendUid}`).child(key).set(true);
    upUnreadCount(friendUid);
  };
};

export const acceptRequest = (uid, friendUid) => {
  return (dispatch) => {
    const promise1 = database()
      .ref(`userFriends/${uid}`)
      .child(friendUid)
      .set('connected');
    const promise2 = database()
      .ref(`userFriends/${friendUid}`)
      .child(uid)
      .set('connected');
    return Promise.all([promise1, promise2]);
  };
};

export const deleteFriend = (uid) => {
  return (dispatch, getState) => {
    const you = getState().profile.profile.uid;
    dispatch(removeFriend(uid));
    return database().ref(`userFriends/${you}`).child(uid).remove();
  };
};

export const fetchFbFriends = (token: string) => {
  return async (dispatch) => {
    const response = await fetch(
      `https://graph.facebook.com/v5.0/me?fields=friends&access_token=${token}`,
    );
    const json = await response.json();
    if (json.friends && json.friends.data) {
      const uids: string[] = await Promise.all(
        json.friends.data.map(async (friend) => {
          const snapshot = await firebase
            .database()
            .ref('fbusers')
            .child(friend.id)
            .once('value');
          return snapshot.val();
        }),
      );
      const validUids = uids.filter((uid) => uid != null);
      return dispatch(fetchUsers(validUids)) || {};
    }
    return {};
  };
};
