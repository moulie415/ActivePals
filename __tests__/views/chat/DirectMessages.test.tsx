import React from 'react';
import { shallow } from 'enzyme';
import DirectMessages from '../../../app/views/chat/DirectMessages';
import mockStore from '../../__mocks__/mockstore';
import friends from '../../fixtures/friends';
import profile from '../../fixtures/profile';
import chats from '../../fixtures/chats';

describe('DirectMessages', () => {
  const initialState = { friends: { friends }, profile: { profile }, chats: { chats } };
  const store = mockStore(initialState);
  describe('Rendering', () => {
    it('should match to snapshot', () => {
      const component = shallow(<DirectMessages store={store} />);
      expect(component).toMatchSnapshot();
    });
  });
});
