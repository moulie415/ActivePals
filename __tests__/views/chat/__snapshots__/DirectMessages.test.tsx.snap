// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DirectMessages Rendering should match to snapshot 1`] = `ShallowWrapper {}`;
